samcli/hook_packages/terraform/hooks.py

```diff
@@ -0,0 +1,344 @@
+    md5 = hashlib.md5()
+    md5.update(bucket.encode())
+    md5.update(key.encode())
+    return md5.hexdigest()
+
+
+def _map_s3_sources_to_functions(s3_hash_to_source: Dict[str, str], cfn_resources: dict) -> None:
+    """
+    Maps the source property of terraform AWS S3 object resources into the the Code property of
+    CloudFormation AWS Lambda Function resources
+
+    Parameters
+    ----------
+    s3_hash_to_source: Dict[str, str]
+        Mapping of S3 object hash to S3 object source
+    cfn_resources: dict
+        CloudFormation resources
+    """
+    for _, resource in cfn_resources.items():
+        if resource.get("Type") == CFN_AWS_LAMBDA_FUNCTION:
+            code = resource.get("Properties").get("Code")
+            bucket = code.get("S3Bucket")
+            key = code.get("S3Key")
+            if bucket and key:
+                obj_hash = _get_s3_object_hash(bucket, key)
+                source = s3_hash_to_source.get(obj_hash)
+                if source:
+                    code["ZipFile"] = source
+                    del code["S3Bucket"]
+                    del code["S3Key"]
+
+
+class TerraformHooks(Hooks):
+    """
+    Hooks implementation for Terraform
+    """
+
+    def __init__(self):
+        pass
+
+    def prepare(self, params: dict) -> dict:
+        """
+        Prepares a terraform application for use with the SAM CLI
+
+        Parameters
+        ----------
+        params: dict
+            Parameters of the IaC application
+
+        Returns
+        -------
+        dict
+            information of the generated metadata files
+        """
+        try:
+            output_dir_path = params.get("OutputDirPath")
+
+            # initialize terraform application
+            run(["terraform", "init", "-upgrade"], check=True, capture_output=True)
+
+            # get json output of terraform plan
+            with NamedTemporaryFile() as temp_file:
+                run(["terraform", "plan", "-out", temp_file.name], check=True, capture_output=True)
+                result = run(["terraform", "show", "-json", temp_file.name], check=True, capture_output=True)
+            tf_json = json.loads(result.stdout)
+
+            # convert terraform to cloudformation
+            cfn_dict = _translate_to_cfn(tf_json)
+
+            # store in supplied output dir
+            if not os.path.exists(output_dir_path):
+                os.mkdir(output_dir_path)
+            metadataFilePath = os.path.join(output_dir_path, "template.json")
+            with open(metadataFilePath, "w+") as metadata_file:
+                json.dump(cfn_dict, metadata_file)
+
+            return {"iac_applications": {"MainApplication": {"metadata_file": metadataFilePath}}}
+
+        except (CalledProcessError, OSError) as e:
+            # one of the subprocess.run calls resulted in non-zero exit code or some OS error
+            raise PrepareHookException("There was an error while preparing the Terraform application.") from e
```

Comment:
@hawflau and I briefly discussed two different ways of handling errors:
```
except SomeInternalException as e:
    raise ExceptionUnderstoodByWarpper(e)
```
```
except SomeInternalException as e:
    return {
        "status": "ERROR",
        "error_message": str(e)
    }
```

We wanted to bring this for discussion in this PR